python/src/tstools/errorFunc.py

```diff
@@ -0,0 +1,951 @@
+
+    elif mode == ifio.THREE_DIM:
+
+        for i, param in enumerate(paramMap[0]):
+            
+            paramMap_i = [paramMap[0][i],paramMap[1][i]]
+
+            x1_partial = xHatPartial( paramMap_i, tsObs, X1, 
+                                      mdlFileHat, brkFileHat)
+            
+            x2_partial = xHatPartial( paramMap_i, tsObs, X2, 
+                                      mdlFileHat, brkFileHat)
+            
+            x3_partial = xHatPartial( paramMap_i, tsObs, X3, 
+                                      mdlFileHat, brkFileHat)
+
+            deltaX1 = tsObs.pos[0] - tsHat.pos[0]
+            
+            deltaX2 = tsObs.pos[1] - tsHat.pos[1]
+            
+            deltaX3 = tsObs.pos[2] - tsHat.pos[2]
+
+            gradVec[i] = 2*np.sum(
+                           (deltaX1*x1_partial)/(tsObs.sig[0]**2)
+                          +(deltaX2*x2_partial)/(tsObs.sig[1]**2)
+                          +(deltaX3*x3_partial)/(tsObs.sig[2]**2))
+
+    return gradVec
+
+########################################################################
+def xHatPartial( param, tsObs, component, mdlFile, brkFile):
+
+    """
+    Compute the partial derivative of x-hat w.r.t. the given parameter 
+    (param) at the epoch of interest (time) for the component (x1,x2,x3)
+    of interest.
+
+    Input(s):
+    param       - list with 2 integers constructed from the paramMap.
+                  first item in list is the value from the 1st column of 
+                  paramMap for the parameter of interest, the 2nd item is 
+                  the value from the 2nd column of paramMap.
+    tsObs       - TimeSeries object with observation data    
+    component   - use constants X1, X2, or X3 to pass function which 
+                  component the parital is being computed for
+    mdlFile     - MdlFile object containing the current value of the 
+                  non-break parameters being estimated.
+    brkFile     - BrkFile object containing the current values of the
+                  Tsbreak model parameters being estimated.
+
+    Output(s):
+    partial     - the partial derivative of x-hat w.r.t. the parameter of
+                  interest evaluated at the time of interest and with 
+                  the current model parameters for the component of 
+                  interest.
+    """
+    
+    # get 1D numpy time array from tsObs and reference to the 
+    # reference epoch in mdlFile object
+    time = tsObs.time - mdlFile.re    
+
+    # parameters from the mdlFile will have a zero-th index of 0
+    if param[0] == params.NON_BRK:
+
+        if param[1] == params.DC_X1:
+
+            if component == X1:
+            
+                partial = np.array([1.]*time.shape[0])
+
+            elif component == X2:
+
+                partial = np.array([0.]*time.shape[0])
+
+            elif component == X3:
+
+                partial = np.array([0.]*time.shape[0])
```

Comment:
This is equivalent to:

```python
if component == X1:
  partial = np.array([1.]*time.shape[0])
elif component == X2 or component == X3:
  partial = np.array([0.]*time.shape[0])
```

There's lots of examples like that in the following code that could be made more succinct. This is an _almost_ 750 line if statement, which might be a bit intense to figure out down the road.

You should add a TODO to your project board to revisit the approach here and make it something a bit more maintainable.